### DISCUSSION TITLE
SessionManager::new_session() method

### DISCUSSION THREAD

**Discussion by Mr-Leshiy** [paritytech/substrate#7801] [Q&A] [answered]
I am trying to figure out how the SessionManager works with the new_session method.
Maybe some can bring some insights into this.
So, we have new_session method we the following signature.
```Rust
fn new_session(new_index: SessionIndex) -> Option<Vec<ValidatorId>>;
```
The argument new_index starts from the value '0' for the first session in the Substrate.
The question is, Does this argument new_index corresponds to the current blockchain height, which is already finalized ? I mean if we have blockchain with the height 100, the next 'new_index' value which will be provided to the new_session method will be 101, or something else ?
I have found that new_session function executed before blocks start to be processed here with the new_index values '0' and '1',
https://github.com/paritytech/substrate/blob/37a7727353f9bac1258acf59cd81688354c9120d/frame/session/src/lib.rs#L450 .
 And after that, each execution of the new_session() method corresponds to the execution of the on_finalize() function.
 ```Rust
  fn on_finalize(block_number: T::BlockNumber)
 ```
 So, I have assumed the following approximate formula for the 'new_index'.
 ```
 block_number = (new_index - 2) * Period
 ```
 Period -  Ends the session after a fixed period of blocks.
 https://github.com/paritytech/substrate/blob/37a7727353f9bac1258acf59cd81688354c9120d/frame/session/src/lib.rs#L139

Am I right with this formula ?
Do we have some guaranties for the 'new_index' values ?

**Comment by gui1117** [comment, marked as answer]
The formula doesn't include the `Offset` information of the periodic session implementation. Formula needs to include it in its general case.
(also formula is only correct for periodic session, if the chain uses Babe to end session then the formula doesn't stand anymore).

Generally speaking SessionManager is used like this:
at the end of session 10: `SessionManager::end_session(10)` is called, `SessionManager::start_session(11)` is called, and `SessionManager::new_session(12)` is called and if it result in a new validator set then it is stored in order to be used at session 12.

But I'm surprised why you need to compute the block number from the session index ? as said previously depending on the `ShouldEndSession` implementation there might be no easy formula (e.g. Babe)

**Reply by Mr-Leshiy** [reply]
Thanks for the answer !
I am trying to modify the validator set, and the information about the validator update initially I am trying to proceed during the block processing. 
Here it is the scheme I am trying to implement
```
[block 10] - [validator set 10]
[block 11] - [validator set 11]
[block 12] - [validator set 12]
```
So in the 'block 10' we got 'validator set 10' and this validator set should be used for the 'block 11' and so on.
